### PR TITLE
Python3 compability, also fixed #17

### LINKED ISSUE #17: Oauth2GoogleClient object has no attribute cache_key_base
I'm trying to port gcs-oauth2-boto-plugin to make it Python2+Python3 compatible, but ran into this issue when cloning from master:

>  File "/enplore-code/environments/enplore-worker-aircraft-dfdr/site-packages/boto/connection.py", line 377, in authorize
>     connection._auth_handler.add_auth(self, **kwargs)
>   File "/enplore-code/environments/enplore-worker-aircraft-dfdr/site-packages/gcs_oauth2_boto_plugin/oauth2_plugin.py", line 69, in add_auth
>     self.oauth2_client.GetAuthorizationHeader()
>   File "/enplore-code/environments/enplore-worker-aircraft-dfdr/site-packages/gcs_oauth2_boto_plugin/oauth2_client.py", line 345, in GetAuthorizationHeader
>     return 'Bearer %s' % self.GetAccessToken().token
>   File "/enplore-code/environments/enplore-worker-aircraft-dfdr/site-packages/gcs_oauth2_boto_plugin/oauth2_client.py", line 310, in GetAccessToken
>     cache_key = self.CacheKey()
>   File "/enplore-code/environments/enplore-worker-aircraft-dfdr/site-packages/gcs_oauth2_boto_plugin/oauth2_client.py", line 335, in CacheKey
>     h.update(self.cache_key_base)
> AttributeError: 'Oauth2GoogleClient' object has no attribute 'cache_key_base'

I'm using a boto config, with a gs service key file:
`gs_service_key_file = /etc/enplore/vagrant-etc/testdemo-s652y8s5.json`

The issue doesn't exist when installing from pip but appeared in the January 27 commit (https://github.com/GoogleCloudPlatform/gcs-oauth2-boto-plugin/commit/53df987472d2440cd8661f2370f33f5e44aa115e)

I've spent a few hours on this, and would love some help. From what I can tell:

**oauth2_helper.py**

> ```
> if json_key:
>   return oauth2_client.Oauth2GoogleClient(
>     oauth2client.client.GoogleCredentials.from_stream(private_key_filename)
>   )
> ```

Doesn't return a oauth2_client with `cache_key_base` attribute which is later needed for `CacheKey()`:

> ```
>  def CacheKey(self):
> """Computes a cache key.
> The cache key is computed as the SHA1 hash of the refresh token for user
> accounts, or the hash of the gs_service_client_id for service accounts,
> which satisfies the FileSystemTokenCache requirement that cache keys do not
> leak information about token values.
> Returns:
>   A hash key.
> """
>     h = sha1()
>     h.update(self.cache_key_base.encode('utf-8'))
>     return h.hexdigest()
> ```

However, I think there is a bunch of stuff missing on the `Oauth2GoogleClient` which is available on `OAuth2ServiceAccountClient`, like `access_token_cache`, `FetchAccessToken` etc.

I'm gonna do a pull request on my own branch and link to it from here, hopefully someone other then me can tell if I re-implemented it correctly (I've tested on py2 and py3 and it works on both)

### PR DESCRIPTION
**Pull request by bivald**
Gives the boto plugin python3 compatibility, as well as solve issue #17 

All tests works, as well as my testing in real life, we're using this in production.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -43,7 +43,7 @@
 
 setup(
     name='gcs-oauth2-boto-plugin',
-    version='1.9',
+    version='1.9.1',
```
Comment (vapier) [inline setup.py]: you shouldn't mess with this file.  just land your fixes and leave release stuff to the maintainers.

File: gcs_oauth2_boto_plugin/oauth2_helper.py [hunk 2]
```diff
@@ -168,7 +168,12 @@ def OAuth2ApprovalFlow(client, scopes, launch_browser=False):
   # Short delay; webbrowser.open on linux insists on printing out a message
   # which we don't want to run into the prompt for the auth code.
   time.sleep(2)
-  code = input('Enter the authorization code: ')
+
+  try:
```
Comment (vapier) [inline gcs_oauth2_boto_plugin/oauth2_helper.py]: you're basically reverting changes made in the previous 2to3 commit.  you should delete/squash those into one change so the history isn't full of noise.

**Comment by vapier** [comment]
instead of doing a merge, you should do a rebase

**Comment by bivald** [comment]
If I do, are Google interested in merging the pull request?

**Comment by vapier** [comment]
i can't speak for the maintainers of this repo, but the PR as-is shouldn't be merged imo

**Comment by bivald** [comment]
That's fine, the reason I'm asking is that since gcs-oauth2-boto-plugin chances rapidly any PR is soon outdated. Thus I'm happy to do a rebase and do it "properly" but it's unnecessary work to do it after every commit on master branch. So if any maintainer is looking, if you are interested in merging a py3 PR ping me on this thread and I'll fix everything.

**Comment by bivald** [comment]
@jonparrott If you feel that this looks OK in terms of the actual code (when viewed on https://github.com/GoogleCloudPlatform/gcs-oauth2-boto-plugin/pull/18/files) I can create a new patch based on a diff between my fork and master as a single commit. 

**Background**:
I appologize that this pull request is a complete mess, I've spend an hour trying to rebase but rebasing through merges in different points in time is above my git skills I'm afraid. As a background this code was not intended as a pull request initially, but was simply used as an internal repo for python3 support with default .boto on GCE but has grown since. It should fix #10 #17 #21 (thanks to @romange). 

These commits adds Python3 support, as well as continues on @romange work regarding oauth2client (added missing wrappers for when using json files as credentials)

I've confirmed that this works (simply using boto to download a file from GCS) manually on: 

| Python version  | Credentials | oauth2client==1.5.2  | oauth2client==2.1.0  |
|---|---|---|---|
| 2  | Default Service Account (on GCE)  | works  | works  |
| 3  |  Default Service Account (on GCE) | works  | works  |
| pypy2  | Default Service Account (on GCE)  |  works | works  |
| 2  | JSON credentials in boto.cfg  | works  | works  |
| 3  | JSON credentials in boto.cfg  | works  |  works |
| pypy2  | JSON credentials in boto.cfg  |  works |  works |

**Comment by jonparrott** [comment]
@bivald don't sweat git issues, we can now squash on merge so it's no big deal.

**Comment by jonparrott** [comment]
Looks okay to me, but @thobrla should be the one to merge.

File: gcs_oauth2_boto_plugin/oauth2_client.py [hunk 3]
```diff
@@ -438,7 +470,7 @@ def GetCredentials(self):
 # be refactored into oauth2client directly in a way that allows for setting of
 # user agent and scopes. https://github.com/google/oauth2client/issues/164
 # pylint: disable=protected-access
-class ServiceAccountCredentials(service_account._ServiceAccountCredentials):
+class ServiceAccountCredentials(ServiceAccountCredentials):
```
Comment (thobrla) [inline gcs_oauth2_boto_plugin/oauth2_client.py]: In the case of oauth2client >=2.0.0, I think we should rely on their existing classes (which support from_json, to_json).

**Comment by thobrla** [comment]
In general we're amenable to this pull request, but I have a competing commit that adds 1.5.2 and >=2.0.0 compatibility.

In general, I think it would be best to separate out that compatibility and Python 3 compatibility into separate comments.  Would you be willing to create a commit for just the Python 3 portion based on my oauth2client cross-compatibility commit?  I intend to push that one within the next few days.

**Comment by bivald** [comment]
Absolutely, no problem. Merge that one and I can add py3 as a clean commit. If it's romanges, could you also look at:

https://github.com/GoogleCloudPlatform/gcs-oauth2-boto-plugin/commit/2a06a44af0e32899fe383d8425fc03868751657c

https://github.com/GoogleCloudPlatform/gcs-oauth2-boto-plugin/commit/44f4a14997bb85d142d9176c1df34b1571105eca

Which complements that one for use with json files (Python 2)

**Comment by thobrla** [comment]
Commit is now pushed, and 1.13 released: https://github.com/GoogleCloudPlatform/gcs-oauth2-boto-plugin/commit/6492539fc8eef91ca815e1710fffb1600756530e

File: gcs_oauth2_boto_plugin/test_oauth2_client.py [hunk 4]
```diff
@@ -14,16 +14,20 @@
...
+
 
 import datetime
 import httplib2
 import logging
-import mox
+try:
+  from mox3 import mox
+except ImportError:
+  import mox as mox
```
Comment (vapier) [inline gcs_oauth2_boto_plugin/test_oauth2_client.py]: the `as mox` is redundant

**Comment by njainua** [comment]
Any updates on this PR?

**Comment by thobrla** [comment]
Would be fine to merge a Python 3 support commit now that oauth2client 2 is supported, but this PR has not yet been updated.

**Comment by njainua** [comment]
@thobrla When do you see this merge happening?

**Comment by thobrla** [comment]
As soon as the PR author (or another author) resolves the conflicts.  It's unlikely the gsutil team will port this module themselves anytime soon.

**Comment by bivald** [comment]
(I'm the author) I'm afraid I haven't had the time to resolve the conflicts, and solved my problem my moving on from boto to using the gcloud-python cloud storage instead.

**Comment by thobrla** [comment]
@njainua it looks like the original author won't be updating this, so I'm going to close this.  If you'd like to submit a pull request of your own based on this one we'd happily accept; otherwise we probably won't port this until gsutil supports Python 3.

**Comment by vapier** [comment]
for posterity, commit 9425f1dd80716888ae2a95b4d572de068f2f9c14 and such should address py3 compat

**Pull request closed without merging**